cdc/sinkv2/ddlsink/cloudstorage/cloud_storage_ddl_sink.go

```diff
@@ -0,0 +1,135 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstorage
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tiflow/cdc/contextutil"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sinkv2/ddlsink"
+	"github.com/pingcap/tiflow/cdc/sinkv2/metrics"
+	"github.com/pingcap/tiflow/pkg/sink"
+	"github.com/pingcap/tiflow/pkg/sink/cloudstorage"
+)
+
+// Assert DDLEventSink implementation
+var _ ddlsink.DDLEventSink = (*ddlSink)(nil)
+
+type ddlSink struct {
+	// id indicates which changefeed this sink belongs to.
+	id model.ChangeFeedID
+	// statistic is used to record the DDL metrics
+	statistics *metrics.Statistics
+	storage    storage.ExternalStorage
+	tables     *model.TableSet
+}
+
+// NewCloudStorageDDLSink creates a ddl sink for cloud storage.
+func NewCloudStorageDDLSink(ctx context.Context, sinkURI *url.URL) (*ddlSink, error) {
+	// parse backend storage from sinkURI
+	bs, err := storage.ParseBackend(sinkURI.String(), &storage.BackendOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// create an external storage.
+	storage, err := storage.New(ctx, bs, nil)
+	if err != nil {
+		return nil, err
+	}
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
+
+	d := &ddlSink{
+		id:         changefeedID,
+		storage:    storage,
+		tables:     model.NewTableSet(),
+		statistics: metrics.NewStatistics(ctx, sink.TxnSink),
+	}
+
+	return d, nil
+}
+
+func (d *ddlSink) generateSchemaPath(def cloudstorage.TableDef) string {
+	return fmt.Sprintf("%s/%s/%d/schema.json", def.Schema, def.Table, def.Version)
+}
+
+func (d *ddlSink) WriteDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	var def cloudstorage.TableDef
+
+	def.FromTableInfo(ddl.TableInfo)
+	encodedDef, err := json.MarshalIndent(def, "", "    ")
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	path := d.generateSchemaPath(def)
+	err = d.statistics.RecordDDLExecution(func() error {
+		err1 := d.storage.WriteFile(ctx, path, encodedDef)
+		if err1 != nil {
+			return err1
+		}
+
+		d.tables.Add(ddl.TableInfo.ID)
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func (d *ddlSink) WriteCheckpointTs(ctx context.Context,
+	ts uint64, tables []*model.TableInfo,
+) error {
+	for _, table := range tables {
+		ok := d.tables.Contain(table.ID)
+		// if table is not cached before, then create the corresponding
+		// schema.json file anyway.
+		if !ok {
+			var def cloudstorage.TableDef
+			def.FromTableInfo(table)
+
+			encodedDef, err := json.MarshalIndent(def, "", "    ")
```

Comment:
why use `MarshalIndent` but not `Marshal`, I think the latter one should be much faster and cheaper?

Reply:
At first, the DDL event is emitted not that frequently. Secondly, we expect to provide a human-readable json file to users. So I think use `MarshalIndent` is acceptable here.